rfcs/2020-09-28-cloudwatch-metrics.md

```diff
@@ -0,0 +1,179 @@
+
+```toml
+[sources.cloudwatch]
+  type = "aws_cloudwatch_metrics"
+  assume_role = "arn:aws:iam::123456789098:role/my_role" # optional, no default
+  endpoint = "127.0.0.0:5000/path/to/service" # optional, no default, relevant when region = ""
+  region = "us-east-1" # required, required when endpoint = ""
+
+  period = "300" # period (s) to aggregate metrics over, optional, can be overridden at metric level, default 300
+  delay = "300" # delay collection by value (s), used to avoid collecting data that has not fully been processed by CloudWatch, optional, default 300
+  interval = "300" # interval to scrape metrics; should be a multiple of "period"; default 300
+
+  metrics_refresh_interval = 900 # interval to refresh available metrics for queried namespaces if globbing or all metrics are used; default 900
+
+  # Request
+  request.in_flight_limit = 25 # optional, default, requests
+  request.rate_limit_duration_secs = 1 # optional, default, seconds
+  request.rate_limit_num = 25 # optional, default
+  request.retry_attempts = 18446744073709551615 # optional, default
+  request.retry_initial_backoff_secs = 1 # optional, default, seconds
+  request.retry_max_duration_secs = 10 # optional, default, seconds
+  request.timeout_secs = 30 # optional, default, seconds
+
+  [[sources.cloudwatch.metrics]]
+    namespace = "AWS/EC2" # required
+    names = ["EBSReadOps", "EBSReadBytes", "Network*"] # optional; defaults to all metrics in namespace, ["*"], (refreshed on interval); supports globbing
+    dimensions.InstanceId = "i-05517fbc2e6124dfb" # optional; supported dimensions differ by namespace and metric; supports globbing
+    statistics = [ "average", "sum", "minimum", "maximum", "sample_count" ] # statistics to collect; can also contain extended statistics like p99; default shown
+    period = 300 # period (s) to aggregate metrics over; defaults to top-level `period` setting; top-level interval should be a multiple of this and any other defined periods
+```
+
+We could alternatively model dimensions as another table:
+
+```toml
+[[sources.cloudwatch.metrics.dimensions]]
+  key = "InstanceId"
+  value = "i-05517fbc2e6124dfb"
+```
+
+To support:
+
+* globbing metric names
+* fetching all metrics for a namespace
+* globbing of dimension values
+
+We will refresh and cache the available metrics on an interval
+(`metrics_refresh_interval`) for namespaces appearing in the configuration.
+
+A wildcard dimension value (`"*"`) will pull and publish discrete metrics for
+each dimension value. This is different than omitting the dimension altogether
+which would instead aggregate across the values for the dimension.
+
+Note that dimensions must match the dimensions that the metrics were published
+with. For example, if the metric was published two dimensions,
+availability-zone and load balancer name, then both dimensions must be
+specified to retrieve the metric. CloudWatch won't automatically aggregate
+across the other if only one is specified.
```

Comment:
Can we warn when this happens or does CW not indicate when this happens?

Reply:
It just doesn't return any metrics. We could warn that no metrics were collected for a given combination, but that could also occur if, for example, they are collecting metrics from an EC2 instance that goes away.

We might be able to cross-reference with the cached metrics from `ListMetrics` to see if any metrics match the given combination though. I'll add that to future work if that's OK; I view it as a nice-to-have.

Reply:
I agree that checking the `ListMetrics` results would be a nice solution, but definitely a "nice to have". Anyone that wants to pull metrics from CW is likely to be aware of this characteristic of CW.